### PR TITLE
JCRVLT-575: handle symlinks in homeDir path

### PR DESCRIPTION
**Pull request by reschke**
Functional changes besides logging:

1. do not call "createDirectories" when the path already exists (this avoids a failure when the last path segment is a symlink)
2. when walking the homeDir, follow symlinks

### CODE REVIEW

**Review by DominikSuess — APPROVED** [review]
lgtm - validated changes under the conditions that provoked failures before (registry from a symlinked locatioN)

**Comment by kwin** [comment]
@reschke Can you add a test leveraging https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createSymbolicLink-java.nio.file.Path-java.nio.file.Path-java.nio.file.attribute.FileAttribute...- to make sure it doesn't regress in the future?

File: vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/registry/impl/FSInstallStateCache.java [hunk 1]
```diff
@@ -53,10 +58,14 @@
     private Map<Path, PackageId> pathIdMapping = new ConcurrentHashMap<>();
 
     private final Path homeDir;
-    
+
     public FSInstallStateCache(Path homeDir) throws IOException {
         this.homeDir = homeDir;
-        Files.createDirectories(homeDir);
+        log.debug("checking for presence of {} - exists {} - isDirectory {}", homeDir, Files.exists(homeDir), Files.isDirectory(homeDir));
+        if (Files.notExists(homeDir)) {
```
Comment (kwin) [inline vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/registry/impl/FSInstallStateCache.java:65]: What happens if the path `homeDir` is a file and not a directory?

**Review by cziegeler — APPROVED** [review]

File: vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/registry/impl/FSInstallStateCache.java (hunk 1 above)
Comment (reschke) [reply vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/registry/impl/FSInstallStateCache.java:65]: Ok, in that case we should fail. Will do.

**Comment by reschke** [comment]
> @reschke Can you add a test leveraging https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createSymbolicLink-java.nio.file.Path-java.nio.file.Path-java.nio.file.attribute.FileAttribute...- to make sure it doesn't regress in the future?

Is there an existing test class that I can extend? (Note that this is a regression from 3.4.0 which works with symlinks just fine, and we really really want to finally upgrade)

**Comment by kwin** [comment]
> Is there an existing test class that I can extend?

There is org.apache.jackrabbit.vault.packaging.registry.impl.FSInstallStateTest (pretty obvious, right?)

> and we really really want to finally upgrade

Yes, I am not saying that this blocks the merge, but it has regressed before because there were no test covering that. Let's do better this time!

**Comment by reschke** [comment]
@kwin - please see https://github.com/apache/jackrabbit-filevault/pull/188/commits/b88b2838cf8a53ba83d20c51e821ca3931f5476b

**Review by kwin — APPROVED** [review]
LGTM, thanks

**Pull request merged**